src/core/cuda/stream_pool.cu

```diff
@@ -0,0 +1,72 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "core/cuda/cuda_help.h"
+#include "core/cuda/stream_pool.h"
+#include "core/runtime/runtime.h"
+
+namespace legate {
+namespace cuda {
+
+StreamView::~StreamView()
+{
+  if (valid_ && Core::synchronize_stream_view) {
+#ifdef DEBUG_LEGATE
+    CHECK_CUDA_STREAM(stream_);
+#else
+    CHECK_CUDA(cudaStreamSynchronize(stream_));
+#endif
+  }
```

Comment:
Can you explain the intent behind `synchronize_stream_view` mode? Based on how `StreamView` is used in cuNumeric, it looks like we will be creating one view at the beginning of the task and automatically destroying it at exit. With this assumption it looks like `synchronize_stream_view` mode is meant to ensure that the stream is automatically sync'd at exit, but we already get this through Realm's automatic context sync. The debug case does more, by also checking for runtime errors through `cudaPeekAtLastError`, but in that case why not do that always when debug mode is active, instead of also requiring `synchronize_stream_view`?

All in all, why not this instead?
```
 StreamView::~StreamView()
 {
 #ifdef DEBUG_LEGATE
   if (valid_) {
     CHECK_CUDA_STREAM(stream_);
   }
 #endif
}
```

Reply:
This PR is actually a prerequisite for my follow-up PR on the NVTX integration. The goal is to have NVTX ranges capture the entire kernel execution and that needs a stream synchronize before each NVTX range is closed. So the RAII on streams would do  that work automatically, as long as the gpu tasks are using stream views "correctly". Alternatively, we could do this in Realm; basically we can close the outstanding NVTX range in the context synchronizer. I don't want to take that route yet, so this is what I came up with.

Note that we don't necessarily want to always synchronize at the end of a scope, and we normally use NVTX ranges for profiling in release mode. So I think we do need `synchornize_stream_view`. A side benefit of being able to toggle the behavior from the outside is that we can easily isolate the cost of host-side execution.

Technically, I can call device synchronize instead before closing each range. I actually did that in the first attempt, but there were some NVTX ranges that report much longer time than their corresponding kernel execution time, and this made me suspect there's some driver-level serialization when multiple threads are calling device synchronize at the same time. (That said, calling stream synchronize concurrently from multiple threads isn't free of such serialization...)